common/serde-reflection/src/lib.rs

```diff
@@ -5,13 +5,110 @@
 
 //! # Serde Reflection (experimental)
 //!
-//! This crate provides a way to extract IDL-like format descriptions for values
-//! that implement the Serialize trait of Serde.
+//! This crate provides a way to extract IDL-like format descriptions for Rust containers
+//! that implement the Serialize and/or Deserialize trait(s) of Serde.
+//!
+//! ```rust
+//! # use libra_serde_reflection::*;
+//! # use serde::{Deserialize, Serialize};
+//! # #[derive(Serialize, PartialEq, Eq, Debug, Clone)]
+//! struct Name(String);
+//! #
+//! # impl<'de> Deserialize<'de> for Name {
+//! #     fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+//! #     where
+//! #         D: ::serde::Deserializer<'de>,
+//! #     {
+//! #         // Make sure to wrap our value in a container with the same name
+//! #         // as the original type.
+//! #         # [derive(Deserialize)]
+//! #         # [serde(rename = "Name")]
+//! #         struct InternalValue(String);
+//! #
+//! #         let value = InternalValue::deserialize(deserializer)?.0;
+//! #         // Enforce some custom invariant
+//! #         if value.len() >= 2 && value.chars().all(char::is_alphabetic) {
+//! #             Ok(Name(value))
+//! #         } else {
+//! #             Err(<D::Error as ::serde::de::Error>::custom(format!(
+//! #                 "Invalid name {}",
+//! #                 value
+//! #             )))
+//! #         }
+//! #     }
+//! # }
+//!
+//! # #[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+//! enum Person {
+//!     NickName(Name),
+//!     FullName { first: Name, last: Name },
+//! }
+//!
+//! # fn main() -> Result<(), Error> {
+//! // Start a session to trace formats.
+//! let mut tracer = Tracer::new(/* is_human_readable */ false);
+//!
+//! // For every type (here `Name`), if a user-defined implementation of `Deserialize`
+//! // exists that performs custom validation checks, start by tracing the serialization
+//! // of a valid value of this type.
+//! let bob = Name("Bob".into());
+//! tracer.trace_value(&bob)?;
+//!
+//! // Then, we may trace deserialization for top-level types (here `Person`) and their
+//! // dependencies as follows. This phase will use Rust values previously recorded
+//! // during serialization in order to pass custom checks.
+//! // TODO: Currently, recorded values are only used for `NewTypeStruct`s.
+//! let format = tracer.trace_type::<Person>()?;
+//! assert_eq!(format, Format::TypeName("Person".into()));
```

Comment:
You can call `trace_type` on any number of types before calling `registery()` right?

Reply:
That's right.